src/Microsoft.AspNetCore.JsonPatch/Internal/DynamicObjectAdapter.cs

```diff
@@ -97,15 +97,15 @@ public class ExpandoObjectAdapter : IAdapter
             out object nextTarget,
             out string errorMessage)
         {
-            var expandoObject = target as ExpandoObject;
-            if (expandoObject == null)
+            var dynamicObject = target as DynamicObject;
+            if (dynamicObject == null)
             {
                 errorMessage = null;
                 nextTarget = null;
                 return false;
             }
 
-            var dictionary = (IDictionary<string, object>)expandoObject;
+            var dictionary = (IDictionary<string, object>)dynamicObject;
```

Comment:
This isn't a good assumption. We need to use `IDynamicMetaObjectProvider` directly.